PassphrasePrompt updates and additions for showing a UI around the passphrase and handle displaying a resulting status message

dritan-x opened the pull request:
- Added title and content header and Cancel and Next Prompt actions UI.
- Added the maximum number of attempts (3) for guessing the passphrase, aka: running the Passphrase prompt.
- Added a Result Status message that is shown at the end of Presentment. The message is shown for 1 second then fades out. If the Presentment was successful it shows a green "Success!" message with 40% background opacity, otherwise it shows "Unable to complete Presentment" message with 40% opacity.
- Added isFixedLength() to PassphraseConstraints and updated all files that use this functionality.
- Made PresentationActivity transparent again.
- Added ImeAction param to PassphraseEntryField to change the action button as needed by the situation, such as when dealing with fixed length passphrases where the action button is not observed.

Test: Manually tested on Android by performing substantial amounts of
      QR and NFC engagements.
Test: ./gradlew check
Test: ./gradlew connectedCheck


#### Success Result Status Message
https://github.com/openwallet-foundation-labs/identity-credential/assets/1477025/a047e71e-3227-4285-adcf-82f3ace10276


#### Unable to complete Presentment
https://github.com/openwallet-foundation-labs/identity-credential/assets/1477025/5323324c-f112-462f-a3b4-216b0491e267


#### PIN with min 3 and max of 6 digits
https://github.com/openwallet-foundation-labs/identity-credential/assets/1477025/0da13e6a-e439-47fb-a396-eef172700577


#### PIN with 6 fixed digits
https://github.com/openwallet-foundation-labs/identity-credential/assets/1477025/42586548-4728-402a-a923-95fbd110ecfb

dritan-x commented:
@davidz25 latest changes as of this message:

Next button present and disabled

https://github.com/openwallet-foundation-labs/identity-credential/assets/1477025/91e2f3bf-5561-4957-bef4-c0ba3f3421f2


Next button enabled after reaching the minimum (3) 


https://github.com/openwallet-foundation-labs/identity-credential/assets/1477025/39d590b8-e4c9-4d1b-93fe-e7fc6b1d11a2

davidz25 commented:
Thanks, what I asked for was for the Next button to always be shown if the PIN length isn't fixed. But it looks like it's now also shown for fixed-digit PINs?

dritan-x commented:
@davidz25 I've pushed changes to the PR addressing your comments + work for handling light and dark theme of device.

https://github.com/user-attachments/assets/80bb9543-7076-4f10-b40c-5656f1615853

https://github.com/user-attachments/assets/39797854-cec7-46e7-be13-243b3bef248a

The background colors and font colors are from the `MaterialTheme` color scheme of the app. We can easily change those to something better suited.

I've spent sooo much time trying to adjust the Passphrase Prompt to show above the keyboard when either collapsing and expanding the keyboard or changing the keyboard altogether. It's proven to be really tricky and I want it to be in a better place.

davidz25 commented:
Please squash the commits together.

davidz25 commented on wallet/src/main/java/com/android/identity_credential/wallet/PresentationActivity.kt:
```diff
@@ -136,16 +175,114 @@ class PresentationActivity : FragmentActivity() {
+        ERROR(R.string.presentation_result_error_message, R.drawable.error_48x48),
+
+        // For init of Flow
+        INIT(0, 0)
```
Perhaps `UNSET` is a better name.

davidz25 commented on wallet/src/main/java/com/android/identity_credential/wallet/PresentationActivity.kt:
```diff
@@ -136,16 +175,114 @@ class PresentationActivity : FragmentActivity() {
+        val coroutineScope = rememberCoroutineScope() // Get the coroutine scope
+
+        var visible by remember { mutableStateOf(true) }
+        if (resultStatus != ResultStatus.INIT) {
```
Cleaner to just invert the logic and return. That way we don't have to indent the stuff below.

davidz25 commented on wallet/src/main/java/com/android/identity_credential/wallet/PresentationActivity.kt:
```diff
@@ -221,12 +357,14 @@ class PresentationActivity : FragmentActivity() {
                             // send the response with all the Document CBOR bytes that succeeded Presentation Flows
                             sendResponseToDevice(deviceResponseGenerator.generate())
+                            _resultStatus.value = ResultStatus.SUCCESS
                         } catch (exception: Exception) {
```
`Throwable`, not `Exception`. Also, use `e`, not `exception` as the parameter name.

davidz25 commented on wallet/src/main/java/com/android/identity_credential/wallet/presentation/PresentmentFlow.kt:
```diff
@@ -136,20 +137,35 @@ suspend fun showPresentmentFlow(
+                    val content =
+                        if (constraints.requireNumerical) activity.resources.getString(R.string.passphrase_prompt_pin_content) else activity.resources.getString(
+                            R.string.passphrase_prompt_passphrase_content
+                        )
```
Need braces/newlines around all the if/else statements. e.g.
```
if (condition) {
  stmt
} else {
  stmt
}
```

davidz25 commented on wallet/src/main/java/com/android/identity_credential/wallet/presentation/PresentmentFlow.kt:
```diff
@@ -136,20 +137,35 @@ suspend fun showPresentmentFlow(
-                    check(passphrase.isNotEmpty()) { "[Passphrase Unsuccessful]" }
-                    // use the passphrase that the user entered to create the KeyUnlockData
+                    // user cancelled the prompt or no text was provided
+                    check(!passphrase.isNullOrEmpty()) { "[Passphrase Unsuccessful]" }
```
What's up with the brackets in error messages? Please remove those (here and elsewhere).

davidz25 commented on wallet/src/main/java/com/android/identity_credential/wallet/ui/prompt/passphrase/PassphraseEntryField.kt:
```diff
@@ -42,28 +42,33 @@ import kotlinx.coroutines.launch
+ * @param imeAction optional, the icon to show for the keyboard action button. If parameter is not
+ *      provided or is [null] (default case), then it uses [ImeAction.Done].
+ *      Note: only the [ImeAction.Done] is recognized to have a task that is executed when the
+ *      user taps on the keyboard's action button.
```
Why do we need this? Want to avoid as this will make moving `PassphraseEntryField` into a Compose Multiplatform library more tricky.

dritan-x replied:
This is an optional parameter for specifying the icon of the action button on the keyboard, if it's not specified (such as in existing composables using `PassphraseEntryField`) `it resorts to use the default "Done".

Regarding Kotlin Multiplatform, this parameter will probably end up being very important to implementing the iOS side of the keyboard and customization of the "action button" for when there's a fixed length passphrase and if the user sees the action button as a "checkmark" they will be misled into thinking they can complete the passphrase prompt by tapping on the keyboard button, but the button won't do anything so the user will think our app is broken and yield to loss of trust -- versus, if we show an arrow (or maybe even remove the action button altogether on iOS - not possible on android) then the user will continue to trust the prompt and the prompt will behave how the user expects it to behave.

davidz25 commented:
Why is Cancel not all the way to the left?

<img width="298" alt="Screenshot 2024-07-15 at 11 39 15 AM" src="https://github.com/user-attachments/assets/aac50cb4-6d1f-472b-a221-7dc12c38ea49">

davidz25 commented:
I think what I asked for was a green icon with text underneath it... this looks like a white icon on a dark green background:
- The icon needs to be bigger and rendered in bright green
- The next needs to be way smaller so "The information was shared" fits in single line

Ditto for the error icon.

<img width="291" alt="Screenshot 2024-07-15 at 11 40 14 AM" src="https://github.com/user-attachments/assets/84ee2495-471b-4e98-9907-52d461c1ab1a">

davidz25 commented:
Where is the cancel button?

<img width="299" alt="Screenshot 2024-07-15 at 11 43 49 AM" src="https://github.com/user-attachments/assets/d43231f3-1c02-4436-87e9-bb55e6221bca">

davidz25 commented:
> I've spent sooo much time trying to adjust the Passphrase Prompt to show above the keyboard when either collapsing and expanding the keyboard or changing the keyboard altogether. It's proven to be really tricky and I want it to be in a better place.

Would it help if the passphrase prompt was always fullscreen?

davidz25 commented on wallet/src/main/res/values/themes.xml:
```diff
@@ -2,4 +2,12 @@
+
+    <style name="BottomSheetTransparentBackground" parent="Widget.Design.BottomSheet.Modal">
+        <item name="android:background">@android:color/transparent</item>
+    </style>
```
Why make this transparent? It looks really weird that way...

davidz25 replied:
I think this can be removed?

dritan-x replied:
It's defined in  the `onCreate` of the Dialog

```
 override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setStyle(STYLE_NORMAL, R.style.PassphrasePromptBottomSheetTheme)
    }
```

And without it it won't follow the dark theme
https://github.com/user-attachments/assets/e89e9699-28be-47ca-b8d1-189edd4a9c4a

With it
![dark-theme](https://github.com/user-attachments/assets/1494ba4a-6920-4691-bfb5-7231ef92e3ae). 

Which brings us to then question the theme colors of the app and adjust them appropriately since we are using theme colors everywhere

dritan-x commented:
Sorry it took a big longer to get this one out (removal of code cannot really be rushed and UI always takes the longest). All comments are welcome and encouraged 🙂

### Latest updates to PR:

#### Full screen Passphrase Prompt
<img width="431" alt="Screenshot 2024-07-18 at 2 57 31 AM" src="https://github.com/user-attachments/assets/45d7a1b5-1676-45df-994b-208bf8d206bb">

#### Result Status Success
<img width="297" alt="result-status-success" src="https://github.com/user-attachments/assets/a9627fb6-c61c-4449-8ddd-7b42ef1a3f34">

#### Result Status Error
<img width="292" alt="result-status-error" src="https://github.com/user-attachments/assets/1d9854ab-9fe9-48d8-9182-fab1f29e0939">

#### Video - changing keyboards

https://github.com/user-attachments/assets/b29dac21-685f-4710-b01e-dc46145cba85

davidz25 commented:
> full-screen-passphrase-prompt.mov

Thanks for the video, this is helpful... in the first presentation it looks like you cancel the dialog and in response it shows "An error occurred". That seems wrong, the "An error occurred" should only be shown if an actual error occurs, not if the user decides to cancel the transaction...

davidz25 commented on identity/src/commonMain/kotlin/com/android/identity/securearea/PassphraseConstraints.kt:
```diff
@@ -30,4 +30,17 @@ data class PassphraseConstraints(
+     *
+     * @return [true] if this object's properties [minLength] and [maxLength] are the same, else [false]
+     */
+    fun isFixedLength() = minLength == maxLength
```
Cc @sorotokin that we should fix `processor` as per the above.

davidz25 commented on wallet/src/main/java/com/android/identity_credential/wallet/ui/prompt/passphrase/PassphrasePrompt.kt:
```diff
@@ -22,23 +68,32 @@ import kotlin.coroutines.resume
+            // if dialog is dismissed during anytime other than producing the passphrase
+            onDismissDialog = {
+                // resume with an exception since the dialog isn't successful
+                continuation.resumeWithException(IllegalStateException("Prompt was dismissed"))
```
Should probably not throw since we don't want to display "An error occurred" when the user decides to cancel the prompt...

dritan-x replied:
Initially I had a sealed class with 3 types, Success, Cancel, and Error that can carry payloads, however during Consent Prompt development you said it was unnecessary and to just throw an exception if user taps on the Cancel button. I followed suit with this prompt.. should I revert both prompts to use a sealed class? or just this prompt? or should I filter through `IllegalStateException` messages that are thrown at the end for keywords such as "User canceled" (taking care to update the message wherever its thrown)? I vote for the sealed class approach to unify a common mechanism for communicating results from each Prompt.

dritan-x replied:
But also I think with the current infra in place it would take more effort than its worth to add the sealed class response (vs having the infra to begin with and grow it organically) and the fastest solution for what we have would be to filter through the Exception messages for predefined text "User Canceled" exception messages.

davidz25 commented on wallet/src/main/java/com/android/identity_credential/wallet/ui/prompt/passphrase/PassphrasePrompt.kt:
```diff
@@ -47,18 +102,71 @@ suspend fun showPassphrasePrompt(
+ * @param onPassphraseResult callback issued from this dialog fragment when the user enters a passphrase
  * @extends [BottomSheetDialogFragment] that can create the Fragment's contents via Composition.
  */
 class PassphrasePrompt(
```
I think this can be a private class, no?

dritan-x replied:
Whenever it's set to be a private class, it compiles but trying to show the prompt yields to

```
java.lang.IllegalStateException: Fragment ...wallet.ui.prompt.passphrase.PassphrasePrompt 
must be a public static class to be  properly recreated from instance state.
```

The same for ConsentPrompt 
`ConsentPrompt must be a public static class to be  properly recreated from instance state.`

davidz25 commented on wallet/src/main/java/com/android/identity_credential/wallet/ui/prompt/passphrase/PassphrasePrompt.kt:
```diff
@@ -69,18 +177,130 @@ class PassphrasePrompt(
+                // rather than a checkmark icon (since with fixed length, the user should not
+                // have the option to "complete" the entry, rather it is completed inherently
+                // when the required number of characters has been entered)
+                val imeAction = if (constraints.isFixedLength()) ImeAction.Go else ImeAction.Done
```
Why `ImeAction.Go` and not `ImeAction.None`? (I'm assuming `.None` will not show anything)

dritan-x replied:
Right first thought was to use `None` but that doesn't do anything - it resorts to the default case which is worse than the "right arrow" of the "Go" `ImeAction`

davidz25 commented on wallet/src/main/java/com/android/identity_credential/wallet/ui/prompt/passphrase/PassphrasePrompt.kt:
```diff
@@ -47,18 +102,71 @@ suspend fun showPassphrasePrompt(
+        /* height of the UI of the Passphrase Prompt (header, text field entry, buttons on bottom)
+           used in calculations for keyboard height and determining placement of UI Prompt so it's
+           always above the keyboard. */
+        const val PROMPT_UI_HEIGHT_DP = 297
```
Looks like this is now unused?

davidz25 commented on wallet/src/main/java/com/android/identity_credential/wallet/ui/prompt/passphrase/PassphrasePrompt.kt:
```diff
@@ -47,18 +102,71 @@ suspend fun showPassphrasePrompt(
+    val keyboardHeightFlow: StateFlow<Int> = _keyboardHeightFlow
+
+    // updated from
+    var currentScreenHeight = 0
```
This too?

dritan-x commented:
Amongst other fixes, the latest changes to the PR addresses cases when the user cancels the prompt

https://github.com/user-attachments/assets/4f004273-cc83-4771-8e98-3f735163c96f

davidz25 reviewed (approved):
OK, let's land this... have a couple of more comments but I'll follow up with them in a PR on top!
